### DISCUSSION TITLE
non blocking delay

### DISCUSSION THREAD

**Discussion by pike2k** [earlephilhower/arduino-pico#2501] [Q&A] [answered]
I am a beginner and I am learning. 
Arduino supports millis() for non blocking delays. I understand Pico lacks this support. 
Is there a similar alternative, maybe something incorporated in this core?

**Comment by FeuerSturm** [comment, marked as answer]
You can use millis() without any problems with this core.

Maybe you have a logical error in your code,
so I'd double check that.

You can show the part of your code if you don't find the mistake.

**Reply by pike2k** [reply]
So I need a delay timer. I'm trying this code, but it results in 0 delay before it performs 

Simplified code below

unsigned long previousMillis = 0;
unsigned long currentMillis = 0;
const long interval = 3000;

void setup () {
  unsigned long previousMillis = millis();
}

void shut_down() {
  unsigned long currentMillis = millis();
  if (currentMillis - previousMillis >= interval) {
        digitalWrite(OUT_OFF, LOW);                      < this happens but with 0 delay
  }
}

void loop() {
  en_status = digitalRead(IN_ENABLE);
  if (en_status == 0)
      {
  shut_down();
      }
  else if (en_status == 1)
  {
...
}

**Reply by dquadros** [reply]
Notice that you declared previousMillis  as a local variable in setup(). loop() uses the global previousMillis variable, that was initialized to zero. Maybe that is the cause of the perceived zero delay,

I tried the following code and it worked OK:

```
unsigned long previousMillis;
const unsigned long interval = 3000;

void setup() {
  previousMillis = millis();
  while(!Serial) {
  }
  Serial.begin(115200);
  Serial.print("Previous: ");
  Serial.println(previousMillis);
}

void loop() {
  unsigned long currentMillis = millis();
  if (currentMillis - previousMillis >= interval) {
    Serial.print("*ding* ");
    Serial.println(currentMillis);
    previousMillis = currentMillis;
  }
}

```

**Reply by pike2k** [reply]
Thanks for the explanation that this core indeed support millis (when this is uncertain everything is uncertain!).

Your code indeed works so I am currently investigating and simplifying my own code (disabling libraries). The only major difference with my delay is that it is not repeating like yours is.

**Reply by pike2k** [reply]
Maybe it is time to explain what I am trying to do. When shut_down()  is called, I want a set delay before toggling the pin OUT_OFF.  So I guess the previousMillis set in setup is of no use for this. Back to the drawing board :)

**Reply by Andy2No** [reply]
It's called the "scope" of a variable.  When you declare a variable inside a function (e.g. in setup() or shut_down()) that variable is only seen inside the function, and is separate to any global variable with the same name.

You had:

```
unsigned long previousMillis = 0;
unsigned long currentMillis = 0;
const long interval = 3000;

void setup () {
unsigned long previousMillis = millis();
}

void shut_down() {
unsigned long currentMillis = millis();
// ...
```

So, you declared global variables at the top:
unsigned long previousMillis = 0;
unsigned long currentMillis = 0;

Then in setup(), you declared a new one, with the same name as one of the global ones, and assigned a value to it:

```
void setup () {
unsigned long previousMillis = millis();
}
```

That variable only did anything within setup(), and ceased to exist when the CPU left setup(), so that value was lost and was never stored in the global variable of the same name.

What you meant to do was this:

```
void setup () {
  previousMillis = millis();
}
```

I.e. just assign to the global variable, not declare a new one and assign to that.

The same sort of thing happened in shut_down():

```
void shut_down() {
unsigned long currentMillis = millis();
```

Again, this declared a new variable, separate from the global one of the same name, and that variable only existed within shut_down().  Since you wanted to use the global variable, that should have been:

```
void shut_down() {
  currentMillis = millis();
```

**Reply by Andy2No** [reply]
If you want the program to just pause for a short while before doing the next thing, you can just use delay() with a number of milliseconds, e.g.:

```
 delay(100);  
```

That makes it pause for 1/10 of a second (100ms).

**Comment by pike2k** [comment]
Thank for this discussion, it got me on the right track.
